_posts/2019-03-28-smart-proxy-capabilities-explained.md

```diff
@@ -0,0 +1,205 @@
+
+```json
+{
+  "facts": {
+    "http_enabled": false,
+    "https_enabled": false,
+    "settings": {},
+    "state": "disabled",
+    "capabilities": []
+  },
+  "dns": {
+    "http_enabled": false,
+    "https_enabled": false,
+    "settings": {
+      "use_provider": null
+    },
+    "state": "disabled",
+    "capabilities": []
+  },
+  "templates": {
+    "http_enabled": true,
+    "https_enabled": true,
+    "settings": {},
+    "state": "running",
+    "capabilities": []
+  },
+  "tftp": {
+    "http_enabled": false,
+    "https_enabled": true,
+    "settings": {
+      "tftp_servername": null
+    },
+    "state": "running",
+    "capabilities": []
+  },
+  "dhcp": {
+    "http_enabled": false,
+    "https_enabled": false,
+    "settings": {
+      "use_provider": null
+    },
+    "state": "disabled",
+    "capabilities": []
+  },
+  "puppetca": {
+    "http_enabled": false,
+    "https_enabled": false,
+    "settings": {
+      "use_provider": null
+    },
+    "state": "failed",
+    "capabilities": []
+  },
+  "puppet": {
+    "http_enabled": false,
+    "https_enabled": true,
+    "settings": {
+      "puppet_url": "https://puppet.example.com:8140",
+      "use_provider": [
+        "puppet_proxy_puppet_api"
+      ]
+    },
+    "state": "running",
+    "capabilities": []
+  },
+  "bmc": {
+    "http_enabled": false,
+    "https_enabled": false,
+    "settings": {},
+    "state": "disabled",
+    "capabilities": []
+  },
+  "realm": {
+    "http_enabled": false,
+    "https_enabled": false,
+    "settings": {
+      "use_provider": null
+    },
+    "state": "disabled",
+    "capabilities": []
+  },
+  "logs": {
+    "http_enabled": false,
+    "https_enabled": true,
+    "settings": {},
+    "state": "running",
+    "capabilities": []
+  },
+  "httpboot": {
+    "http_enabled": false,
+    "https_enabled": false,
+    "settings": {},
+    "state": "disabled",
+    "capabilities": []
+  }
+}
+```
+
+As can be seen, it is now a mapping of the feature name to another map representing the feauture's state.
+
+`http_enabled` and `https_enabled` are the result of `enabled` in the feature's setting file. That can be `true`, `false`, `http` or `https`. The installer configures most features using `https` for increased security, but some features need to listen on HTTP. `Templates` and `HTTPBoot` are examples of this because connecting clients wouldn't be able to verify HTTPS or not deal with it at all.
+
+Where we previously relied on the installer to verify, the new datastructure exposes `state`. The current known values for this are `uninitialized`, `loaded`, `starting`, `running`, `disabled` and `failed`. As can be seen in the example, it's mostly `disabled` or `running`. The interesting one is `failed`. Now the Foreman application itself can detect misconfigurations although this hasn't been implemented. In the above output the puppetca feature has failed which is also why it wasn't present in the classic registration.
```

Comment:
Is there a guidance or a definition of each state? What's the difference between loaded and running? OT: Is there some way to signal there's non-fatal error?

Reply:
Good point. This should be formalized but currently it's just the internal state.

Reply:
So can we add some one sentence description of states here?